cpp/src/parquet/column_writer.cc

```diff
@@ -534,22 +534,16 @@ void ColumnWriter::FlushBufferedDataPages() {
 template <typename Type>
 TypedColumnWriter<Type>::TypedColumnWriter(ColumnChunkMetaDataBuilder* metadata,
                                            std::unique_ptr<PageWriter> pager,
+                                           const bool use_dictionary,
                                            Encoding::type encoding,
                                            const WriterProperties* properties)
-    : ColumnWriter(metadata, std::move(pager),
-                   (encoding == Encoding::PLAIN_DICTIONARY ||
-                    encoding == Encoding::RLE_DICTIONARY),
-                   encoding, properties) {
-  switch (encoding) {
-    case Encoding::PLAIN:
-      current_encoder_.reset(new PlainEncoder<Type>(descr_, properties->memory_pool()));
-      break;
-    case Encoding::PLAIN_DICTIONARY:
-    case Encoding::RLE_DICTIONARY:
-      current_encoder_.reset(new DictEncoder<Type>(descr_, properties->memory_pool()));
-      break;
-    default:
-      ParquetException::NYI("Selected encoding is not supported");
+    : ColumnWriter(metadata, std::move(pager), use_dictionary, encoding, properties) {
+  if (use_dictionary) {
```

Comment:
Why did you change the old logic from the switch to the extra `use_dictionary` flag?

Reply:
This wasn't obvious to me either.  For the version 2.0 code path it seems that we need to instantiate `ColumnWriter` with the `RLE_DICTIONARY` encoding but previously it was using the `PLAIN` encoding as defined here in arrow/cpp/src/parquet/properties.h:

```
  inline Encoding::type dictionary_page_encoding() const {
    if (parquet_version_ == ParquetVersion::PARQUET_1_0) {
      return Encoding::PLAIN_DICTIONARY;
    } else {
      return Encoding::PLAIN;
    }
  }
```

To fix this, I replaced the call to `dictionary_page_encoding()` with `dictionary_index_encoding()` in `ColumnWriter::Make'.  It seemed clearer to me to also compute the `use_dictionary` option there too and pass it down the stack.  Let me know if you have any suggestions to make this clearer?

Reply:
Thanks for clarifying, that makes a bit more sense to me now.